DatadogCore/Sources/Core/Upload/BackgroundTaskCoordinator.swift

```diff
@@ -60,4 +60,38 @@ internal class UIKitBackgroundTaskCoordinator: BackgroundTaskCoordinator {
         self.currentTaskId = nil
     }
 }
+
+/// Bridge protocol that matches `UIApplication` interface for background tasks. Allows easier testablity.
+internal protocol ProcessInfoActivityCoordinator {
+    func beginActivity(options: ProcessInfo.ActivityOptions, reason: String) -> any NSObjectProtocol
+    func endActivity(_ activity: any NSObjectProtocol)
+}
+
+extension ProcessInfo: ProcessInfoActivityCoordinator {}
+
+internal class ExtensionBackgroundTaskCoordinator: BackgroundTaskCoordinator {
+    private let processInfo: ProcessInfoActivityCoordinator
+
+    @ReadWriteLock
+    private var currentActivity: NSObjectProtocol?
+
+    internal init(
+        processInfo: ProcessInfoActivityCoordinator = ProcessInfo()
+    ) {
+        self.processInfo = processInfo
+    }
+
+    internal func beginBackgroundTask() {
+        endBackgroundTask()
+        currentActivity = processInfo.beginActivity(options: [.background], reason: "Datadog SDK background upload")
+    }
```

Comment:
question/ Because `beginBackgroundTask()` [is called](https://github.com/DataDog/dd-sdk-ios/blob/ad20f2d4ba50de8f01af7ffcbd18348ebe9fbf5b/DatadogCore/Sources/Core/Upload/DataUploadWorker.swift#L80C49-L80C68) right before the SDK starts uploading data, how does it correspond to the [explanation of `0xdead10cc`](https://developer.apple.com/documentation/xcode/sigkill)?

> (...) Make this request **well before starting to write to the file** in order to complete those operations and relinquish the lock before the app suspends.

Apple suggestion is to wrap "file **write**" operations with `.beginActivity()` and `.endActivity()`. What we seem to do instead is wrapping data uploads. Considering that [upload starts](https://github.com/DataDog/dd-sdk-ios/blob/ad20f2d4ba50de8f01af7ffcbd18348ebe9fbf5b/DatadogCore/Sources/Core/Upload/DataUploadWorker.swift#L80-L81) after the [file was **read**](https://github.com/DataDog/dd-sdk-ios/blob/ad20f2d4ba50de8f01af7ffcbd18348ebe9fbf5b/DatadogCore/Sources/Core/Upload/DataUploadWorker.swift#L77) and it has nothing to do with file writes (which are decoupled SDK routine), I'm not seeing clearly how this solves the `0xdead10cc` problem.

Shouldn't we instead use `.beginActivity()` and `.endActivity()` when writing to files, instead of when performing data upload 🤔💭?